### ISSUE TITLE
NWConnection and its NWParameters should be exposed for a channel

### ISSUE THREAD

**Issue by PasiSalenius** [apple/swift-nio-transport-services#88]
### Expected behavior
This part in the Swift NIO Transport Services README made me assume that I could access the NWConnection and NWParameters of the established channel for a NIOTSListenerBootstrap or a NIOTSConnectionBootstrap.

> In addition to providing first-class support for Apple platforms, NIO Transport Services takes advantage of the richer API of Network.framework to provide more insight into the behaviour of the network than is normally available to NIO applications.

Basically, I'm looking to get statistics of a connection after its completion, things like NWParameters, NWProtocolMetadata, EstablishmentReports, DataTransferReports etc.

### Actual behavior
I cannot find any way to get to these connection details. The class NIOTSConnectionChannel seems to be internal, which prevents me from casting from Channel to that. Basically, all I'm left with is the same amount of connection info I had without NIO TS.

### If possible, minimal yet complete reproducer code (or URL to code)

```
let clientBootStrap = NIOTSConnectionBootstrap(group: context.eventLoop)
      .channelInitializer { channel in
            // no access to NIOTSConnectionChannel here...
      }
```

### SwiftNIO version/commit hash

swift-nio-transport-services version 1.5.1

### Swift & OS version (output of `swift --version && uname -a`)

Apple Swift version 5.2.2 (swiftlang-1103.0.32.6 clang-1103.0.32.51)

**Comment by Lukasa** [comment]
Yeah, we have thus far not exposed top-level metadata like this. Generally speaking having access to the channel object would not help you either, as it has almost no public API surface area. What we need is some channel options that you can use to retrieve this information.

It seems like we could have a few channel options

- `currentPath`, for accessing the `NWConnection.currentPath`
- `metadata(definition)`, for accessing the metadata for a given `NWProtocol`.
- `establishmentReport`, for accessing the establishment report
- `dataTransferReport`, for initiating a data transfer report.

Are you interested in adding those channel options? We can provide some help so you know where to add them.

**Comment by PasiSalenius** [comment]
Sure, I can give it a shot. It can take a few days before I get to it, though.

**Comment by Lukasa** [comment]
Sure, no rush.

**Comment by PasiSalenius** [comment]
Any suggestions how to best proceed getting the dataTransferReport out of the connection?

The reporting must first be initiated to a `NWConnection.PendingDataTransferReport` and then later collected to the final `NWConnection.DataTransferReport`. Should the `PendingDataTransferReport` be returned via a future when getting the channel option, so the user could then call `collect()` on it later when the necessary data transfers are complete?

I have an implementation almost there, but I cannot figure out what `DispatchQueue` to use in `collect(queue: DispatchQueue, completion: @escaping (NWConnection.DataTransferReport) -> Void)` of `NWConnection.PendingDataTransferReport`, as `.main` does not seem to work here.

Or should the channel option handle the report collection and only expose the final data transfer report to the user?

**Comment by Lukasa** [comment]
> Should the `PendingDataTransferReport` be returned via a future when getting the channel option, so the user could then call `collect()` on it later when the necessary data transfers are complete?

I think this is my favourite option of these.

What makes you think `.main` doesn't work?

**Comment by PasiSalenius** [comment]
Because the print never gets executed, when I do this
```
pendingReport.collect(queue: .main) { report in
     print("data transfer report = \(report)")
}
```

**Comment by Lukasa** [comment]
Are you running this in a simple command line binary?

**Comment by PasiSalenius** [comment]
I tried adding reading of all these new options to this
https://github.com/apple/swift-nio-transport-services/blob/master/Sources/NIOTSHTTPClient/main.swift

I may be just doing something stupid, but I can't get the closure in the `collect()` to execute.

**Comment by Lukasa** [comment]
TL;DR: you need to add a call to `dispatchMain()` instead of `channel.closeFuture.wait()`.

The main dispatch queue is not running at the start of program execution. It needs to be given the opportunity to start running. In your average AppKit or UIKit application this is taken care of before you start the program. However, in command-line binaries that define a `main` function (as Swift ones do) that is not automatic, and you need to arrange to do it yourself.

`dispatchMain` will park the main thread and allow the main queue to begin executing. If that's not called, you will never find blocks enqueued to the main queue executing.
